Consolidate and cross-link kubecf documentation. (#87)
SUSE/kubecf: doc/dev/general.md

# Kubecf deployment via any Kube + Operator/Helm + Kubecf/Helm

The intended audience of this document are developers wishing to
contribute to the Kubecf project.

Here we explain how to deploy Kubecf using:

  - A generic kubernetes cluster.
  - A released cf-operator helm chart.
  - A released kubecf helm chart.

## Kubernetes

> In contrast to other recipes we are not set on using a local

Reviewer: ```suggestion
In contrast to other recipes, we are not set on using a local
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

> cluster. Any kubernetes cluster will do. Assuming that the following

Reviewer: ```suggestion
cluster. Any Kubernetes cluster will do, assuming that the following
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

requirements are met:

>   - Presence of a default storage class (provisioner)

Reviewer: ```suggestion
  - Presence of a default storage class (provisioner).
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

>   - For use with a diego-based kubecf (default) a node OS with XFS

Reviewer: ```suggestion
  - For use with a diego-based kubecf (default), a node OS with XFS
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

    support.

>       - For GKE using the option `--image-type UBUNTU` with the

Reviewer: ```suggestion
      - For GKE, using the option `--image-type UBUNTU` with the
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

        `gcloud beta container` command selects such an OS.

> This can be any of

Reviewer: ```suggestion
This can be any of, but not restricted to:
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

  - GKE ([Notes](../provider/gke.md))
  - AKS
  - EKS

> etc.

Reviewer: ```suggestion
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

> 

Reviewer: ```suggestion
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

Note that how to deploy and tear-down such a cluster is outside of the
scope of this recipe.

## cf-operator

The [cf-operator] is the underlying generic tool to deploy a (modified)
BOSH deployment like Kubecf for use.

[cf-operator]: https://github.com/cloudfoundry-incubator/cf-operator

> It has to be installed in the same kube cluster Kubecf will be deployed to.

Reviewer: ```suggestion
It has to be installed in the same Kubernetes cluster that Kubecf will be deployed to.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

Here we are not using development-specific dependencies like bazel,
but only generic tools, i.e. `kubectl` and `helm`.

[Installing and configuring Helm](helm.md) is the same regardless of
the chosen foundation, and assuming that the cluster does not come

> with helm pre-installed (cluster-side).

Reviewer: ```suggestion
with Helm Tiller pre-installed.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

### Deployment and Tear-down

```shell
helm install --name cf-operator \
     --namespace cfo \
     --set "operator.watchNamespace=kubecf" \
     https://s3.amazonaws.com/cf-operators/helm-charts/cf-operator-v0.4.1%2B92.g77e53fda.tgz
```

> In the example above version 0.4.1 of the operator was used.  Look

Reviewer: ```suggestion
In the example above, version 0.4.1 of the operator was used. Look
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

> into the `cf_operator` section of the toplevel `def.bzl` file to find

Reviewer: ```suggestion
into the `cf_operator` section of the top-level `def.bzl` file to find
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

the version of the operator validated against the current kubecf
master.

Note how the namespace the operator is installed into (`cfo`) differs
from the namespace the operator is watching for deployments (`kubecf`).

> This form of deployment enables restarting the operator, because it is

Reviewer: ```suggestion
This form of deployment enables restarting the operator because it is
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

> not affected by webhooks. It further enables the deletion of the

Reviewer: I'm not sure if this statement still holds. We should instruct people to delete the kubecf helm installation and any PVCs that stayed behind (because the database would get confused using the old PVCs).

Author: @f0rmiga Please verify truthiness and if not, make a ticket to fix.

Kubecf deployment namespace to start from scratch, without redeploying
the operator itself.

Tear-down is done with a standard `helm delete ...` command.

## Kubecf

> With all the prequisites handled by the preceding sections it is now

Reviewer: ```suggestion
With all the prerequisites handled by the preceding sections, it is now
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

possible to build and deploy kubecf itself.

This again uses helm and a released helm chart.

### Deployment and Tear-down

```
helm install --name kubecf \
     --namespace kubecf \
     https://scf-v3.s3.amazonaws.com/scf-3.0.0-82165ef3.tgz \
     --set "system_domain=kubecf.suse.dev"
```

> In this default deployment kubecf is launched without Ingress, and

Reviewer: ```suggestion
In this default deployment, kubecf is launched without Ingress, and it
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

uses the Diego scheduler.

Tear-down is done with a standard `helm delete ...` command.

### Access

To access the cluster after the cf-operator has completed the
deployment and all pods are active invoke:

```sh
cf api --skip-ssl-validation "https://api.$(minikube ip).xip.io"

# Copy the admin cluster password.

> acp=$(kubectl get secret \

Reviewer: ```suggestion
admin_pass=$(kubectl get secret \
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

	      --namespace kubecf kubecf.var-cf-admin-password \
	      -o jsonpath='{.data.password}' \
	      | base64 --decode)

# Use the password from the previous step when requested.

> cf auth -u admin -p "${acp}"

Reviewer: ```suggestion
cf auth -u admin -p "${admin_pass}"
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

```

### Advanced Topics

#### Diego vs Eirini

Diego is the standard scheduler used by kubecf to deploy CF

> applications. Eirini is an alternative talking more directly to the

Reviewer: ```suggestion
applications. Eirini is an alternative that talks more directly to the
```

Reviewer: Consider rephrasing the Eirini sentence. Maybe:
```
Eirini is an alternative to Diego that follows a more Kubernetes native approach, deploying the CF apps directly to a Kubernetes namespace.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

underlying Kube cluster.

> To activate this alternative use the option

Reviewer: ```suggestion
To activate this alternative, use the option
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

`--set features.eirini.enabled=true` when deploying kubecf from its chart.

#### Ingress

> By default the cluster is exposed through its kubernetes services.

Reviewer: ```suggestion
By default, the cluster is exposed through its Kubernetes services.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

> To use the NGINX ingress instead it is necessary to:

Reviewer: ```suggestion
To use the NGINX ingress instead, it is necessary to:
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

>   - Install and configure the Nginx Ingress Controller.

Reviewer: ```suggestion
  - Install and configure the NGINX Ingress Controller.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

>   - Configure kubecf to use ingress.

Reviewer: ```suggestion
  - Configure Kubecf to use the ingress controller.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

This has to happen before deploying kubecf.

##### Installation of the NGINX Ingress Controller

```sh
helm install stable/nginx-ingress \
  --name ingress \
  --namespace ingress
  --set "controller.service.externalIPs={$(minikube ip)}"
```

> The last option above assigns the external IP of the cluster to the

Reviewer: ```suggestion
The last option in the command above assigns the external IP of the cluster to the
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

Ingress Controller service.

##### Configure kubecf

> Use the option `--set features.ingress.enabled=true` when deploying

Reviewer: ```suggestion
Use the Helm option `--set features.ingress.enabled=true` when deploying
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

> kubecf from its chart.

Reviewer: ```suggestion
kubecf.
```

Author: Fixed with commit 9918b20dcb257d8865a289cee9653b710a4662a0.

# Use the password from the previous step when requested.
cf auth -u admin -p "${acp}"
```

### Advanced Topics

#### Diego vs Eirini

Diego is the standard scheduler used by kubecf to deploy CF
applications. Eirini is an alternative talking more directly to the
underlying Kube cluster.

To activate this alternative use the option
`--set features.eirini=true` when deploying kubecf from its chart.

#### Ingress

By default the cluster is exposed through its kubernetes services.

> To use the NGINX ingress instead it is necessary to:

Reviewer: Missing a comma after `instead`.

Author: Fixed with commit 71143eb3b3bdd30e920ee562bdea5c5a326f2d67
